pkg/admission/util.go

```diff
@@ -31,38 +31,50 @@ import (
 	"github.com/apache/yunikorn-k8shim/pkg/log"
 )
 
-func updatePodLabel(pod *v1.Pod, namespace string, generateUniqueAppIds bool, defaultQueueName string) map[string]string {
-	existingLabels := pod.Labels
+func getAnnotationsForApplicationInfoUpdate(pod *v1.Pod, namespace string, generateUniqueAppIds bool, defaultQueueName string) map[string]string {
 	result := make(map[string]string)
-	for k, v := range existingLabels {
+	existingAnnotations := pod.Annotations
+	for k, v := range existingAnnotations {
 		result[k] = v
 	}
-
-	sparkAppID := utils.GetPodLabelValue(pod, constants.SparkLabelAppID)
-	appID := utils.GetPodLabelValue(pod, constants.LabelApplicationID)
-	if sparkAppID == "" && appID == "" {
+	appID := getApplicationIDFromPod(pod)
+	if appID == "" {
 		// if app id not exist, generate one
 		// for each namespace, we group unnamed pods to one single app - if GenerateUniqueAppId is not set
 		// if GenerateUniqueAppId:
 		//		application ID convention: ${NAMESPACE}-${GENERATED_UUID}
 		// else
 		// 		application ID convention: ${AUTO_GEN_PREFIX}-${NAMESPACE}-${AUTO_GEN_SUFFIX}
-		generatedID := generateAppID(namespace, generateUniqueAppIds)
-		result[constants.LabelApplicationID] = generatedID
+		appID = generateAppID(namespace, generateUniqueAppIds)
+		result[constants.AnnotationApplicationID] = appID
 
 		// if we generate an app ID, disable state-aware scheduling for this app
-		if _, ok := existingLabels[constants.LabelDisableStateAware]; !ok {
-			result[constants.LabelDisableStateAware] = "true"
+		disableStateAware := "true"
+		if value := utils.GetPodLabelValue(pod, constants.LabelDisableStateAware); value != "" {
+			disableStateAware = value
 		}
+		result[constants.AnnotationDisableStateAware] = disableStateAware
+	}
+
+	// if app id not in pod annotation, add it
+	if value := utils.GetPodAnnotationValue(pod, constants.AnnotationIgnoreApplication); value == "" {
```

Comment:
@craigcondit The `yunikorn.apache.org/ignore-application` was from PR https://github.com/apache/yunikorn-k8shim/pull/579, but it's not on the [document](https://yunikorn.apache.org/docs/user_guide/labels_and_annotations_in_yunikorn). I'm not sure whether users can set the value?

If not, we don't need to check the value here. If yes, we also need to check whether value can be parsed as bool. Thank you.
cc @chenyulin0719

Reply:
Hi @FrankYang0529, 

Sorry, I just realized that I used the wrong constants here. 
I should change the constant `constants.AnnotationIgnoreApplication -> constants.AnnotationApplicationID`.

We should always add app-id to annotation if it wasn't in pod.  Will fix it.

Reply:
Committed a fixed version b08d5cda0d031e630e48928b4810a752b0b2d670.
Also add more comments/tests to make the logic clearer.